README.md
# Excel-GoalSeek

[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/p4lv/Excel-GoalSeek/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/p4lv/Excel-GoalSeek/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/p4lv/Excel-GoalSeek/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/p4lv/Excel-GoalSeek/?branch=master)
[![Build Status](https://scrutinizer-ci.com/g/p4lv/Excel-GoalSeek/badges/build.png?b=master)](https://scrutinizer-ci.com/g/p4lv/Excel-GoalSeek/build-status/master)
[![Code Intelligence Status](https://scrutinizer-ci.com/g/p4lv/Excel-GoalSeek/badges/code-intelligence.svg?b=master)](https://scrutinizer-ci.com/code-intelligence)

Utility to emulate goalseek function in PHP. This fork was made for very specific project, and there is no promise to keep any BC. 

This library is a fork of [PHP-Excel-GoalSeek](https://github.com/juanjomip/PHP-Excel-GoalSeek)

## Install

Use package from packagist.
```bash
composer require p4/excel-goalseek
```

## Usage

``` php
//Define function which for which the value should be found
$callbackTest = function callbackTest($input) {
    $inputForCallbackTest2 = $input * 8;
    return $inputForCallbackTest2 - 12;
};

//Instantiate goal seek class
$goalseek = new ExcelGoalSeek();
//$goalseek->debug = true;

//I want to know which input needs callbackTest to give me 301
$expected_result = 300;

//Calculate the input to get you goal, with accuracy
$input = $goalseek->calculate($callbackTest, $expected_result, 5);

//Voilá!
echo "\$input: " . $input . "<br />";

//Let's test our input it is close
$actual_result = $callbackTest($input);
//Searched result of function
echo "Searched result of $callbackTest(\$input) = " . $expected_result . "<br />";
//Actual result of function with calculated goalseek
echo "Actual result of $callbackTest(" . $input . ") = " . $actual_result . "<br />";
//If difference is too high, you can improve the class and send me it your modifications ;)
echo "Difference = " . ($actual_result - $expected_result);
```

## Testing

``` bash
$ phpunit
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [David Jiménez](https://github.com/davidjr82)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `Excel-GoalSeek` will be documented in this file

## 2021-09-02

- migrated to different namespace.
- increased min version of php
- update to the latest phpunit

## 2021-10-08

- made function to test against a parameter instead of private method 

tests/PHPExcelGoalSeekTest.php
<?php

namespace davidjr82\PHPExcelGoalSeek\Test;

use davidjr82\PHPExcelGoalSeek\PHPExcelGoalSeek;
use davidjr82\PHPExcelGoalSeek\Exceptions\PHPExcelGoalSeekException;

class GoalSeek extends \davidjr82\PHPExcelGoalSeek\PHPExcelGoalSeek {

    function callbackTest($input) {
        $inputForCallbackTest2 = $input * 8;
        return $this->callbackTest2($inputForCallbackTest2);
    }
    function callbackTest2($input) {
        $solution = $input - 12;
        return $solution;
    }

    function callbackTest3($input) {
        $inputForCallbackTest4 = $input * 8;
        return $this->callbackTest4($inputForCallbackTest4);
    }
    function callbackTest4($input) {
        $solution = pow($input, 2);
        return $solution;
    }

}

class PHPExcelGoalSeekTest extends \PHPUnit_Framework_TestCase
{

    protected $goalseek;

    protected function setUp()
    {
        $this->goalseek = new GoalSeek();
    }

    protected function tearDown()
    {
        unset($this->goalseek);
    }

    public static function provider()
    {
        return array(
            array('callbackTest', 302, 5, 39.25),
            array('callbackTest', 301, 5, 39.125),
            array('callbackTest', 300, 5, 39),
            array('callbackTest', -1, 5, 1.375),
            );
    }

    public static function provider3()
    {
        return array(
            array('callbackTest3', -1, 5, null),
            );
    }


    /**
     * @dataProvider provider
     */
    public function testCalculated($testFunction, $goalseeked, $accuracy, $expected)
    {
        // Assert
        $this->assertEquals($this->goalseek->calculate($testFunction, $goalseeked, $accuracy), $expected);
    }

    /**
     * @dataProvider provider3
     */
    public function testCalculated3($testFunction, $goalseeked, $accuracy, $expected)
    {
        // Assert
        $this->assertEquals($this->goalseek->calculate($testFunction, $goalseeked, $accuracy), $expected);
    }


    public function testCalculatedNewLine()
    {
        // Assert
        $this->assertEquals($this->goalseek->newLine(), '<br />');
    }

    /**
     * @expectedException davidjr82\PHPExcelGoalSeek\Exceptions\PHPExcelGoalSeekException
     */
    public function testException()
    {
        $this->goalseek->calculate('', 0, 0);
        $this->setExpectedException();
    }
}


tests/ExcelGoalSeek/ExcelGoalSeekTest.php
<?php

namespace P4lv\ExcelGoalSeek;

use P4lv\ExcelGoalSeek\Exception\GoalNeverReached;
use PHPUnit\Framework\TestCase;
use Psr\Log\NullLogger;

class ExcelGoalSeekTest extends TestCase
{

    protected $goalseek;

    protected function setUp(): void
    {
        $this->goalseek = new ExcelGoalSeek();
    }

    protected function tearDown(): void
    {
        unset($this->goalseek);
    }

    public static function provider()
    {
        $callbackTest = function($input) {
            $input = $input * 8;
            return $input - 12;
        };


        return [
            [$callbackTest, 302, 5, 39.25],
            [$callbackTest, 301, 5, 39.125],
            [$callbackTest, 300, 5, 39],
            [$callbackTest, -1, 5, 1.375],
        ];
    }

    public static function provider3()
    {
        $callbackTest3 = function ($input) {
            $input = $input * 8;
            return pow($input, 2);
        };

        return [
            [$callbackTest3, -1, 5, null],
        ];
    }


    /**
     * @dataProvider provider
     */
    public function testCalculated($testFunction, $goalseeked, $accuracy, $expected)
    {
        // Assert
        self::assertEquals($this->goalseek->calculate($testFunction, $goalseeked, $accuracy), $expected);
    }

    /**
     * @dataProvider provider
     */
    public function testCalculatedWithLogger($testFunction, $goalseeked, $accuracy, $expected)
    {
        $this->goalseek = new ExcelGoalSeek(new NullLogger());
        self::assertEquals($this->goalseek->calculate($testFunction, $goalseeked, $accuracy), $expected);
    }

    /**
     * @dataProvider provider3
     */
    public function testCalculated3($testFunction, $goalseeked, $accuracy, $expected)
    {
        $this->expectException(GoalNeverReached::class);
        // Assert
        self::assertEquals($this->goalseek->calculate($testFunction, $goalseeked, $accuracy), $expected);
    }


    public function testException()
    {
        $this->expectException(\TypeError::class);
        $this->goalseek->calculate('', 0, 0);
    }
}


src/ExcelGoalSeek/ExcelGoalSeek.php
<?php

namespace P4lv\ExcelGoalSeek;

use P4lv\ExcelGoalSeek\Exception\GoalNeverReached;
use P4lv\ExcelGoalSeek\Exception\GoalReachedNotEnough;
use Psr\Log\LoggerInterface;

class ExcelGoalSeek
{
    /**
     * @var LoggerInterface|null
     */
    private $logger;

    public function __construct(LoggerInterface $logger = null)
    {
        $this->logger = $logger;
    }

    private function debug($message, array $context = []): void
    {
        if ($this->logger instanceof LoggerInterface) {
            $this->logger->debug($message, $context);
        }
    }

    public function calculate(
        callable $function,
        $goal,
        $decimal_places,
        $incremental_modifier = 1,
        $max_loops_round = 0,
        $max_loops_dec = 0,
        $lock_min = ['num' => null, 'goal' => null],
        $lock_max = ['num' => null, 'goal' => null],
        $slope = null,
        $randomized = false,
        $start_from = 0.1
    )
    {
        //If goal found has more than this difference, return null as it is not found
        $maximum_acceptable_difference = 0.1;
        $max_loops_round++;

        $this->checkLimitRestriction($max_loops_round);

        $this->debug(sprintf("Iteration %d; min value = %s; max value = %s; slope %s", $max_loops_round, $lock_min['num'], $lock_max['num'], $slope));

        //If I have the goal  limited to a unit, I seek decimals
        if ($lock_min['num'] !== null && $lock_max['num'] !== null && abs(abs($lock_max['num']) - abs($lock_min['num'])) <= 1) {

            //No decimal , return result
            if ($lock_min['num'] == $lock_max['num']) {
                return $lock_min['num'];
            }

            //Seek decimals
            foreach (range(1, $decimal_places, 1) as $decimal) {
                $decimal_step = 1 / (10 ** $decimal);

                $difference = abs(round(abs($lock_max['num']), $decimal) - round(abs($lock_min['num']), $decimal));

                while ($difference - ($decimal_step / 10) > $decimal_step && $max_loops_dec < (2000 * $decimal_places)) {
                    $max_loops_dec++;

                    $aux_obj_num = round(($lock_min['num'] + $lock_max['num']) / 2, $decimal);
                    $aux_obj = $function($aux_obj_num);

                    $this->debug(sprintf("Decimal iteration %d; min value = %s; max value = %s; value %s", $max_loops_dec, $lock_min['num'], $lock_max['num'], $aux_obj));

                    //Like when I look without decimals
                    [$lock_min, $lock_max] = $this->lookWithoutDecimals($aux_obj, $goal, $aux_obj_num, $lock_min, $lock_max, $slope);
                    //End Like when I look without decimals
                    $difference = abs(round(abs($lock_max['num']), $decimal) - round(abs($lock_min['num']), $decimal));
                }//End while
            }//End foreach


            if ($max_loops_dec > 2000 * $decimal_places) {
                throw new GoalNeverReached('Goal never reached [2000]');
            }

            if (!is_nan($lock_min['goal']) && abs(abs($lock_min['goal']) - abs($goal)) < $maximum_acceptable_difference) {
                return round($lock_min['num'], $decimal_places - 1);
            }

            throw new GoalReachedNotEnough('Goal reached not enough');
        }

        //First iteration, try with zero
        $aux_obj_num = $this->getAuxObjNum($lock_min['num'], $lock_max['num'], $start_from, $incremental_modifier);

        $aux_obj = $function($aux_obj_num);

        $this->debug(sprintf("Testing (with initial value) %s%d with value %s", $aux_obj_num != $start_from ? '' : '(with initial value)', $aux_obj_num, $aux_obj));

        if ($slope === null) {
            $aux_slope = $function($aux_obj_num + 0.1);

            if (is_nan($aux_slope) || is_nan($aux_obj)) {
                $slope = null; //If slope is null
            } elseif ($aux_slope - $aux_obj > 0) {
                $slope = 1;
            } else {
                $slope = -1;
            }
        }

        //Test if formule can give me non valid values, i.e.: sqrt of negative value
        if (!is_nan($aux_obj)) {
            //Is goal without decimals?
            [$lock_min, $lock_max] = $this->lookWithoutDecimals($aux_obj, $goal, $aux_obj_num, $lock_min, $lock_max, $slope);
        } else {
            if (($lock_min['num'] === null && $lock_max['num'] === null) || $randomized) {
                $nuevo_start_from = random_int(-500, 500);

                return $this->calculate($function, $goal, $decimal_places, $incremental_modifier + 1, $max_loops_round, $max_loops_dec, $lock_min, $lock_max, $slope, true, $nuevo_start_from);
            } //First iteration is null

            if ($lock_min['num'] !== null && abs(abs($aux_obj_num) - abs($lock_min['num'])) < 1) {
                $lock_max['num'] = $aux_obj_num;
            }
            if ($lock_max['num'] !== null && abs(abs($aux_obj_num) - abs($lock_max['num'])) < 1) {
                $lock_min['num'] = $aux_obj_num;
            }

            return $this->calculate($function, $goal, $decimal_places, $incremental_modifier + 1, $max_loops_round, $max_loops_dec, $lock_min, $lock_max, $slope, $randomized, $start_from);
        }

        return $this->calculate($function, $goal, $decimal_places, $incremental_modifier, $max_loops_round, $max_loops_dec, $lock_min, $lock_max, $slope, $randomized, $start_from);
    }

    private function lookWithoutDecimals($aux_obj, $goal, $aux_obj_num, $lock_min, $lock_max, $slope): array
    {
        if ($aux_obj == $goal) {
            $lock_min['num'] = $aux_obj_num;
            $lock_min['goal'] = $aux_obj;

            $lock_max['num'] = $aux_obj_num;
            $lock_max['goal'] = $aux_obj;
        }

        $going_up = false;
        if ($aux_obj < $goal) {
            $going_up = true;
        }
        if ($aux_obj > $goal) {
            $going_up = false;
        }
        if ($slope == -1) {
            $going_up = !$going_up;
        }

        if ($going_up) {
            if ($lock_min['num'] !== null && $aux_obj_num < $lock_min['num']) {
                $lock_max['num'] = $lock_min['num'];
                $lock_max['goal'] = $lock_min['goal'];
            }

            $lock_min['num'] = $aux_obj_num;
            $lock_min['goal'] = $aux_obj;
        }

        if (!$going_up) {
            if ($lock_max['num'] !== null && $lock_max['num'] < $aux_obj_num) {
                $lock_min['num'] = $lock_max['num'];
                $lock_min['goal'] = $lock_max['goal'];
            }

            $lock_max['num'] = $aux_obj_num;
            $lock_max['goal'] = $aux_obj;
        }
        return [$lock_min, $lock_max];
    }

    /**
     * @param $lockMinNum
     * @param $lockMaxNum
     * @param $start_from
     * @param $incremental_modifier
     * @return float|int|mixed
     */
    private function getAuxObjNum($lockMinNum, $lockMaxNum, $start_from, $incremental_modifier)
    {
        $aux_obj_num = null;
        if ($lockMinNum === null && $lockMaxNum === null) {
            $aux_obj_num = $start_from;
        } //Lower limit found, searching higher limit with * 10
        elseif ($lockMinNum !== null && $lockMaxNum === null) {
            if ($lockMinNum == $start_from) {
                $aux_obj_num = 1;
            } else {
                $aux_obj_num = $lockMinNum * (10 / $incremental_modifier);
            }
        } //Higher limit found, searching lower limit with * -10
        elseif ($lockMinNum === null && $lockMaxNum !== null) {
            if ($lockMaxNum == $start_from) {
                $aux_obj_num = -1;
            } else {
                $aux_obj_num = $lockMaxNum * (10 / $incremental_modifier);
            }
        } //I have both limits, searching between them without decimals
        elseif ($lockMinNum !== null && $lockMaxNum !== null) {
            $aux_obj_num = round(($lockMinNum + $lockMaxNum) / 2);
        }
        return $aux_obj_num;
    }

    /**
     * @param int $max_loops_round
     */
    private function checkLimitRestriction(int $max_loops_round): void
    {
        if ($max_loops_round > 100) {
            throw new GoalNeverReached();
        }
    }
}


src/PHPExcelGoalSeek.php
<?php

namespace davidjr82\PHPExcelGoalSeek;

use davidjr82\PHPExcelGoalSeek\Exceptions\PHPExcelGoalSeekException;

class PHPExcelGoalSeek
{
    var $debug;

    public function __construct()
    {
        $this->debug = false;
    }

    public function newLine() {
        return '<br />';
    }

    public function calculate($functionGS, $goal, $decimal_places,
      $incremental_modifier = 1, $max_loops_round = 0, $max_loops_dec = 0,
      $lock_min = array('num' => null, 'goal' => null), $lock_max = array('num' => null, 'goal' => null),
      $slope = null, $randomized = false, $start_from = 0.1)
    {
        $debug = $this->debug;

        if (empty($functionGS)) {
            throw new PHPExcelGoalSeekException('Function callback expected');
        }

        $max_loops_round++;

        $maximum_acceptable_difference = 0.1;//If goal found has more than this difference, return null as it is not found

        if ($max_loops_round > 100) {
            if ($debug) {
                echo 'Goal never reached'.$this->newLine();
            }
            return false;
        }

        if ($debug) {
            echo 'Iteration '.$max_loops_round.'; min value = '.$lock_min['num'].'; max value = '.$lock_max['num'].'; slope '.$slope.$this->newLine();
        }

        //If I have the goal  limited to a unit, I seek decimals
        if ($lock_min['num'] !== null && $lock_max['num'] !== null && abs(abs($lock_max['num']) - abs($lock_min['num'])) <= 1) {

            //No decimal , return result
            if ($lock_min['num'] == $lock_max['num']) {
                return $lock_min['num'];
            }

            //Seek decimals
            foreach (range(1, $decimal_places, 1) as $decimal) {
                $decimal_step = 1 / pow(10, $decimal);

                $difference = abs(round(abs($lock_max['num']), $decimal) - round(abs($lock_min['num']), $decimal));

                while ($difference - ($decimal_step / 10) > $decimal_step && $max_loops_dec < (2000 * $decimal_places)) {
                    $max_loops_dec++;

                    $aux_obj_num = round(($lock_min['num'] + $lock_max['num']) / 2, $decimal);
                    $aux_obj = $this->$functionGS($aux_obj_num);

                    if ($debug) {
                        echo 'Decimal iteration '.$max_loops_dec.'; min value = '.$lock_min['num'].'; max value = '.$lock_max['num'].'; value '.$aux_obj.$this->newLine();
                    }

                    //Like when I look without decimals
                    if ($aux_obj == $goal) {
                        $lock_min['num'] = $aux_obj_num;
                        $lock_min['goal'] = $aux_obj;

                        $lock_max['num'] = $aux_obj_num;
                        $lock_max['goal'] = $aux_obj;
                    }

                    $going_up = false;
                    $going_down = false;
                    if ($aux_obj < $goal) {
                        $going_up = true;
                    }
                    if ($aux_obj > $goal) {
                        $going_up = false;
                    }
                    if ($slope == -1) {
                        $going_up = !$going_up;
                    }

                    if ($going_up) {
                        if ($lock_min['num'] !== null && $aux_obj_num < $lock_min['num']) {
                            $lock_max['num'] = $lock_min['num'];
                            $lock_max['goal'] = $lock_min['goal'];
                        }

                        $lock_min['num'] = $aux_obj_num;
                        $lock_min['goal'] = $aux_obj;
                    }

                    if (!$going_up) {
                        if ($lock_max['num'] !== null && $lock_max['num'] < $aux_obj_num) {
                            $lock_min['num'] = $lock_max['num'];
                            $lock_min['goal'] = $lock_max['goal'];
                        }

                        $lock_max['num'] = $aux_obj_num;
                        $lock_max['goal'] = $aux_obj;
                    }
                    //End Like when I look without decimals

                    $difference = abs(round(abs($lock_max['num']), $decimal) - round(abs($lock_min['num']), $decimal));
                }//End while
            }//End foreach


        if ($max_loops_dec > 2000 * $decimal_places) {
            if ($debug) {
                echo 'Goal never reached'.$this->newLine();
            }
            return;
        }

            if (!is_nan($lock_min['goal']) && abs(abs($lock_min['goal']) - abs($goal)) < $maximum_acceptable_difference) {
                return round($lock_min['num'], $decimal_places - 1);
            } else {
                if ($debug) {
                    echo 'Goal reached not enough'.$this->newLine();
                }
                return;
            }
        }

    //First iteration, try with zero
    if ($lock_min['num'] === null && $lock_max['num'] === null) {
        $aux_obj_num = $start_from;
    }
    //Lower limit found, searching higher limit with * 10
    elseif ($lock_min['num'] !== null && $lock_max['num'] === null) {
        if ($lock_min['num'] == $start_from) {
            $aux_obj_num = 1;
        } else {
            $aux_obj_num = $lock_min['num'] * (10 / $incremental_modifier);
        }
    }
    //Higher limit found, searching lower limit with * -10
    elseif ($lock_min['num'] === null && $lock_max['num'] !== null) {
        if ($lock_max['num'] == $start_from) {
            $aux_obj_num = -1;
        } else {
            $aux_obj_num = $lock_max['num'] * (10 / $incremental_modifier);
        }
    }
    //I have both limits, searching between them without decimals
    elseif ($lock_min['num'] !== null && $lock_max['num'] !== null) {
        $aux_obj_num = round(($lock_min['num'] + $lock_max['num']) / 2);
    }

        if ($aux_obj_num != $start_from) {
            $aux_obj = $this->$functionGS($aux_obj_num);
            if ($debug) {
                echo 'Testing '.$aux_obj_num.' with value '.$aux_obj.$this->newLine();
            }
        } else {
            $aux_obj = $this->$functionGS($aux_obj_num);
            if ($debug) {
                echo 'Testing (with initial value) '.$aux_obj_num.' with value '.$aux_obj.$this->newLine();
            }
        }

        if ($slope == null) {
            $aux_slope = $this->$functionGS($aux_obj_num + 0.1);

            if (is_nan($aux_slope) || is_nan($aux_obj)) {
                $slope = null; //If slope is null
            } elseif ($aux_slope - $aux_obj > 0) {
                $slope = 1;
            } else {
                $slope = -1;
            }
        }

        //Test if formule can give me non valid values, i.e.: sqrt of negative value
        if (!is_nan($aux_obj)) {
            //Is goal without decimals?
            if ($aux_obj == $goal) {
                $lock_min['num'] = $aux_obj_num;
                $lock_min['goal'] = $aux_obj;

                $lock_max['num'] = $aux_obj_num;
                $lock_max['goal'] = $aux_obj;
            }

            $going_up = false;
            $going_down = false;
            if ($aux_obj < $goal) {
                $going_up = true;
            }
            if ($aux_obj > $goal) {
                $going_up = false;
            }
            if ($slope == -1) {
                $going_up = !$going_up;
            }

            if ($going_up) {
                if ($lock_min['num'] !== null && $aux_obj_num < $lock_min['num']) {
                    $lock_max['num'] = $lock_min['num'];
                    $lock_max['goal'] = $lock_min['goal'];
                }

                $lock_min['num'] = $aux_obj_num;
                $lock_min['goal'] = $aux_obj;
            }

            if (!$going_up) {
                if ($lock_max['num'] !== null && $lock_max['num'] < $aux_obj_num) {
                    $lock_min['num'] = $lock_max['num'];
                    $lock_min['goal'] = $lock_max['goal'];
                }

                $lock_max['num'] = $aux_obj_num;
                $lock_max['goal'] = $aux_obj;
            }
        } else {
            if (($lock_min['num'] === null && $lock_max['num'] === null) || $randomized) {
                $nuevo_start_from = rand(-500, 500);

                return $this->calculate($functionGS, $goal, $decimal_places, $incremental_modifier + 1, $max_loops_round, $max_loops_dec, $lock_min, $lock_max, $slope, true, $nuevo_start_from, $debug);
            } //First iteration is null

            if ($lock_min['num'] !== null && abs(abs($aux_obj_num) - abs($lock_min['num'])) < 1) {
                $lock_max['num'] = $aux_obj_num;
            }
            if ($lock_max['num'] !== null && abs(abs($aux_obj_num) - abs($lock_max['num'])) < 1) {
                $lock_min['num'] = $aux_obj_num;
            }

            return $this->calculate($functionGS, $goal, $decimal_places, $incremental_modifier + 1, $max_loops_round, $max_loops_dec, $lock_min, $lock_max, $slope, $randomized, $start_from, $debug);
        }

        return $this->calculate($functionGS, $goal, $decimal_places, $incremental_modifier, $max_loops_round, $max_loops_dec, $lock_min, $lock_max, $slope, $randomized, $start_from, $debug);
    }
}
